synapse/res/username_picker/script.js

```diff
@@ -0,0 +1,116 @@
+let inputField = document.getElementById("field-username");
+let inputForm = document.getElementById("form");
+let submitButton = document.getElementById("button-submit");
+let message = document.getElementById("message");
+
+// Remove input field placeholder if the text field is not empty
+let switchClass = function(input) {
+  if (input.value.length > 0) {
+    input.classList.add('has-contents');
+  }
+  else {
+    input.classList.remove('has-contents');
+  }
+};
+
+// Submit username and receive response
+let showMessage = function(messageText) {
+  // Unhide the message text
+  message.classList.remove("hidden");
+
+  message.innerHTML = messageText;
+};
+
+let onResponse = function(response, success) {
+  // Display message
+  showMessage(response);
+
+  if(success) {
+    inputForm.submit();
+    return;
+  }
+
+  // Enable submit button and input field
+  submitButton.classList.remove('button--disabled');
+  submitButton.value = "Submit"
+};
+
+let allowedUsernameCharacters = RegExp("[^a-z0-9\\.\\_\\=\\-\\/]");
+let usernameIsValid = function(username) {
+  return !allowedUsernameCharacters.test(username);
+}
+let allowedCharactersString = "" +
+"lowercase letters, " +
+"digits, " +
+"<code>.</code>, " +
+"<code>_</code>, " +
+"<code>-</code>, " +
+"<code>/</code>, " +
+"<code>=</code>";
+
+let buildQueryString = function(params) {
+    return Object.keys(params)
+        .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(params[k]))
+        .join('&');
+}
+
+let submitUsername = function(username) {
+  if(username.length == 0) {
+    onResponse("Please enter a username.", false);
+    return;
+  }
+  if(!usernameIsValid(username)) {
+    onResponse("Invalid username. Only the following characters are allowed: " + allowedCharactersString, false);
+    return;
+  }
+
+    let check_uri = 'check?' + buildQueryString({"username": username});
+    fetch(check_uri, {
+        "credentials": "include",
```

Comment:
What's the reasoning for this here? It seems like there are no credentials to include? (Or does this mean "include cookies"?)

Reply:
The docs for [fetch](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters) explain that `credentials: "include"` indeed will send the relevant browser cookies as part of the request.

Looks like we're actually pulling from the request's cookies in `AvailabilityCheckResource`, so I would leave this in.

Reply:
I'm pretty sure it should be `same-origin`, not `include`. See https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Sending_a_request_with_credentials_included

Reply:
Yep, `include` was probably just used as a "this works" sort of deal. Super happy to constrain to only what's necessary (and what makes CORS work) :)

Reply:
switched to `same-origin`.